### ISSUE TITLE
question about points order

### ISSUE THREAD

**Issue by scliu2018** [charlesq34/pointnet2#143]
Hi, I have question about the points order. If I shuffle the points order, pointnet++ seems output a very different result. Also, I shuffle the points order of training data and found that the loss cannot drop to an ideal value. I think it might be because FPS algorithm always sample certain point set for a input with certain points order.   So does points order really have influence on pointnet++'s result and should I shuffle the points order during training phase?

**Comment by jackd** [comment]
I spent _a lot_ of time trying to work this out. ~As far as I can tell shuffling the points that are used as inputs to the model won't make a difference - it's permutation-invariant in that respect. Having said that,~ the dataset provided here is not randomly shuffled. The image below is of the number of neighbors in a given radius and the distance to the closest neighbor for a single example, either by randomly sampling 1024 of the 10000 point provided vs taking the first 1024. As you can see, they are very different distributions.

So, if you shuffle then take the first 1024 (equivalent to randomly sampling 1024) you'll get a much less even covering compared to taking the first 1024 points in the original ordering. This particular aspect has nothing to do with the network architecture/properties, but will definitely affect results.

~Inside the network, points are sorted with iterative farthest point, so any initial shuffling will wash out.~

Looking deeper, you'll find the `query_ball` code terminates prematurely if it finds the specified maximum number of neighbors. For each query point it iterates through each point in the cloud in the order provided, making the output highly dependent on the order when this cropping is in effect. That said, the authors preprocess the initial cloud into farthest point sampling order, so the first `k` within the ball will be well distributed  - they won't be the closest `k`.

This preprocessing is not provided in the code (they provide the preprocessed points and the FPS algorithm - they just don't call it on the original point cloud), so if you use your own data, you'll need to  either preprocess it into order or modify the model. Note that beyond the first cloud all clouds are in FPS order already, so this difference will only affect queries on the first (which may be significant with MSG, less so SSG).

![dists](https://user-images.githubusercontent.com/659115/63428149-2f3ae380-c45a-11e9-81c9-7715f5de2fcc.png)

**Comment by zubair1502** [comment]
Hello @jackd, thank you for this detail explanation. as you mentioned that author didn't provide the preprocessing of the data in the code but provide the FPS algorithm. I need to preprocess my data that have variable number of points in each pointcloud. I checked the FPS algorithm that author provide but I don't understand how to apply it on my dataset. 

I have already compiled all the bash files successfully but when I run the tf_sampling code with the argument of my pointcloud, It generate .pkl file. How can I get my resampled cloud, I am not getting it.

Thanks.

**Comment by jackd** [comment]
@zubair1502 It's been a while since I've looked at this, but is it not just `prob_sample` (is this the IFP algorithm? I can't remember...) followed by `gather_point`? If you're using a recent version of tf, I believe `gather_point(values, indices)` is equivalent to `tf.gather(values, indices, batch_dims=1)`, but I stand to be corrected.
